src/litmodels/integrations/checkpoints.py

```diff
@@ -25,6 +28,72 @@
         import pytorch_lightning as pl
 
 
+# Create a singleton upload manager
+@lru_cache(maxsize=None)
+def get_model_manager():
+    """Get or create the singleton upload manager."""
+    return ModelManager()
+
+
+class ModelManager:
+    """Manages uploads and removals with a single queue but separate counters."""
+
+    def __init__(self):
+        """Initialize the ModelManager with a task queue and counters."""
+        self.task_queue = queue.Queue()
+        self.upload_count = 0
+        self.remove_count = 0
+        self._worker = threading.Thread(target=self._worker_loop, daemon=True)
+        self._worker.start()
+
+    def _worker_loop(self):
+        while True:
+            task = self.task_queue.get()
+            if task is None:
+                self.task_queue.task_done()
+                break
+            action, detail = task
+            if action == "upload":
+                registry_name, filepath = detail
+                try:
+                    upload_model(registry_name, filepath)
+                    rank_zero_debug(f"Finished uploading: {filepath}")
+                except Exception as ex:
+                    rank_zero_warn(f"Upload failed {filepath}: {ex}")
+                finally:
+                    self.upload_count -= 1
```

Comment:
Tiny comment on why it is decremented there.

Reply:
just looks cleaner :)